### PR TITLE
Feature/complex path index

### PR DESCRIPTION
**Pull request by jeswr**
See https://github.com/LDflex/LDflex/issues/79

### CODE REVIEW

File: src/JSONLDResolver.js [hunk 1]
```diff
@@ -67,9 +68,36 @@ export default class JSONLDResolver {
...
+      });
+    }
+    catch (e) {
       throw new Error(`The JSON-LD context cannot expand the '${property}' property`);
-    return namedNode(expandedProperty);
+    }
+
+    if (algebra.input.type === 'path') {
+      const query = toSparql(algebra);
+      return query.slice(25, query.length - 7);
```
Comment (RubenVerborgh) [inline src/JSONLDResolver.js:92]: Let's not have magic numbers

**Review by RubenVerborgh — COMMENTED** [review]
Thanks! Let's first agree on syntax in #79, and then look into implementation.

That said, I probably want to keep this out of `JSONLDResolver` but instead have a separate `StarResolver` that takes a `JSONLDResolver` as a constructor argument.

As you indeed notice, there's essentially only additions to the code. So let's wrap those additions in a separately testable unit.

File: src/JSONLDResolver.js (hunk 1 above)
Comment (jeswr) [reply src/JSONLDResolver.js:92]: Would you prefer `.replace(/^SELECT \* WHERE \{ \?s \/, '')`, `.slice('SELECT * WHERE { ?s '.length,)` or something else? Would be cleaner if `toSparql` could be run on `algebra.input`; do you know if there are any internals in that library to do so?

**Comment by jeswr** [comment]
> That said, I probably want to keep this out of `JSONLDResolver` but instead have a separate `StarResolver` that takes a `JSONLDResolver` as a constructor argument.
> 
> As you indeed notice, there's essentially only additions to the code. So let's wrap those additions in a separately testable unit.

Would `ComplexPathResolver` be more suitable given the decision in #79

File: src/JSONLDResolver.js (hunk 1 above)
Comment (RubenVerborgh) [reply src/JSONLDResolver.js:92]: The replace would be nice; then we know if things fail.

**Comment by RubenVerborgh** [comment]
> Would `ComplexPathResolver` be more suitable given the decision in #79

Sounds good to me!

**Comment by jeswr** [comment]
I've refactored the additions complex path resolver, though it requires a few duplicate methods form the `JSONLDResolver`. I'd also note that one could make the resolvers independent if 

(a) the `supports` method could be `async` allowing the following to be the supports method in the `JSONLDResolver`
```ts
  async supports(property) {
    if (typeof property !== 'string') return false;
    // JavaScript requires keys containing colons to be quoted,
    // so prefixed names would need to written as path['foaf:knows'].
    // We thus allow writing path.foaf_knows or path.foaf$knows instead.
    property = property.replace(/^([a-z][a-z0-9]*)[_$]/i, '$1:');

    // Expand the property to a full IRI
    const context = await this._context;
    const expandedProperty = context.expandTerm(property, true);
    return ContextUtil.isValidIri(expandedProperty);
```
or (b) LDflex continues to try the rest of the resolvers if one rejects.

On another note I am using node version 15 so `npm install` has updated the `package-lock.json` to a version 2 lockfile. If this is an issue let me know and I can add the `sparqlalgerbajs` dependency using an older version.

**Comment by RubenVerborgh** [comment]
> it requires a few duplicate methods form the `JSONLDResolver`

How come?

I see we are injecting a constructor; shall we inject an actual resolver directly.
And then access the methods you need from there?

> (a) the `supports` method could be `async`

We can't though; we need to decide synchronously, since JavaScript property resolution is synchronous.

> or (b) LDflex continues to try the rest of the resolvers if one rejects.

We could code it that way.

However, I'd expect your `supports` to do a check to see if any complex paths appear in the property name? So like a regular expression check on star patterns etc.?

**Comment by jeswr** [comment]
> How come?

The logic for `resolve` is the same, they both call `expandProperty` and it is this method is implemented differently between the two classes.

Similarly they share the same `getResultsCache` method, with the `JSON-LD` resolver using single predicates as keys, for the property cache whilst the complex path resolver has keys that are the string representation of a SPARQL path. 

They both also require the context (`ComplexPathResolver` just passes the `prefix`, `namespace` pairs to `sparqlalgebrajs`) so it makes sense that the `extendContext` method works in the same manner.

> However, I'd expect your `supports` to do a check to see if any complex paths appear in the property name? So like a regular expression check on star patterns etc.?

Sure, so we could test to see if the string tests true on any of the following expressions
```ts
// Tests for reverse key at start of string or after '/', '|'
const inverseTest = /(^|[/|])[\^]/
// Tests for length modififier e.g. ex:test*, <http://example.org/test>?, (ex:test)+
const lengthModiferTest = /([a-z:>)])[\*\+\?]/i
// Tests for '\' and '|' operators *in* a path
const pathModifierTest = /([a-z:)>\*\+\?])[|/]([a-z:<(\^])/i
// Tests for '(', '<', at the start of a string and ')', '>' at the end of a string
const braceTest = /((^[(<])|([)>]$))/
```
If we did that and ensured that the `ComplexPathResolver` was tested *before* the JSON-LD resolver, then there would be no need to pass the JSON-LD resolver as an argument to the class (since `sparqlalgebrajs` is handling the processing of prefixes within complex paths). 

In this case I guess the shared `resolve` and `getResultsCache` logic could be placed inside an abstract `PathResolver` class.

**Comment by RubenVerborgh** [comment]
> Similarly they share the same `getResultsCache` method, with the `JSON-LD` resolver using single predicates as keys, for the property cache whilst the complex path resolver has keys that are the string representation of a SPARQL path.

We probably want to abstract that; either by pushing that to a utility that both use, or with a base class.

Or can `JSONLDResolver` be the base class?

> They both also require the context (`ComplexPathResolver` just passes the `prefix`, `namespace` pairs to `sparqlalgebrajs`) so it makes sense that the `extendContext` method works in the same manner.

Idem.

However, the code is sufficiently short, so not a major issue for me.

Perhaps see what you can do, but keep the current state as a savepoint 🙂

> Sure, so we could test to see if the string tests true on any of the following expressions

Excellent—we can add them to one giant OR expression with `|`?

> If we did that and ensured that the `ComplexPathResolver` was tested _before_ the JSON-LD resolver,

We can do that.

> then there would be no need to pass the JSON-LD resolver

Yeah but this bit https://github.com/LDflex/LDflex/pull/80/files#diff-864bc7111617e5a3589ebef530a48e5e3ea3bab395d682f75231ca99234da346R50-R69 really seems like it ought to be called from the JSON-LD resolver. Because if that resolver ever changes (and it will) the exact same logic should be used.

**Comment by jeswr** [comment]
I've now factored most of the logic into the `AbstractPathResolver` class. I've also created a `ContextProvider` class which can be passed as input to separate `PathResolvers` so they can share context updates from `.extendContext` calls if necessary.

> Yeah but this bit https://github.com/LDflex/LDflex/pull/80/files#diff-864bc7111617e5a3589ebef530a48e5e3ea3bab395d682f75231ca99234da346R50-R69 really seems like it ought to be called from the JSON-LD resolver. Because if that resolver ever changes (and it will) the exact same logic should be used.

This is now handled by `.runExpandProperty` in the abstract class and the value is then passed to `.expandProperty` in the concrete classes.

**Comment by RubenVerborgh** [comment]
Thanks a lot, @jeswr!

**Pull request merged**